packages/kotti-ui/source/make-props.test.ts

```diff
@@ -1,51 +1,143 @@
-	prop: PropOptions<unknown, boolean>,
-	successCases: unknown[],
-	failCases: unknown[],
-) => {
-	const { validator } = prop
-
-	if (validator === undefined) {
-		expect(validator).not.toBe(undefined)
-		throw new Error('unexpected undefined validator')
+declare global {
+	namespace jest {
+		interface Matchers<R> {
+			toBeRequired(): R
+			toBeType(expectedPropType: PropType<unknown>): R
+			toDefaultTo(defaultValue: unknown): R
+			toValidate(...cases: unknown[]): R
+		}
 	}
-
-	successCases.forEach((c) => expect(validator(c)).toBe(true))
-	failCases.forEach((c) => expect(validator(c)).toBe(false))
 }
 
+expect.extend({
+	toBeRequired(prop: PropOptions<unknown, boolean>) {
+		const passNoDefault = prop.default === undefined
+		const passRequired = prop.required === true
+
+		return {
+			message: () => {
+				if (!passNoDefault)
+					return `Expected “prop.default” to be “undefined”. Was “${prop.default}”`
+				if (!passRequired)
+					return `Expected “prop.required” to be “true”. Was “${prop.required}”`
+				return 'valid'
+			},
+			pass: passNoDefault && passRequired,
+		}
+	},
+	toBeType(prop: PropOptions<unknown, boolean>, expected: PropType<unknown>) {
+		const actualPropTypes = new Set(castArray(prop.type as PropType<unknown>))
+		const expectedPropTypes = new Set(castArray(expected))
+
+		const pass = isEqual(actualPropTypes, expectedPropTypes)
+
+		return {
+			message: () =>
+				pass
+					? 'valid'
+					: this.utils.printDiffOrStringify(
+							expectedPropTypes,
+							actualPropTypes,
+							'Expected prop.type',
+							'Received prop.type',
+							this.expand,
+					  ),
+			pass,
+		}
+	},
+	toDefaultTo(prop: PropOptions<unknown, boolean>, expectedDefault: unknown) {
+		if (typeof prop.default !== 'function')
+			return {
+				message: () => 'Expected “prop.default” to be a function',
+				pass: false,
+			}
+
+		const actualDefaultValue = prop.default()
+		const passDefault = isEqual(actualDefaultValue, expectedDefault)
+		const passNoRequired = prop.required === undefined
+
+		return {
+			message: () => {
+				if (!passDefault)
+					return this.utils.printDiffOrStringify(
+						expectedDefault,
+						actualDefaultValue,
+						'Expected prop.default',
+						'Received prop.default',
+						this.expand,
+					)
+
+				if (!passNoRequired)
+					return this.utils.printDiffOrStringify(
+						undefined,
+						prop.required,
+						'Expected prop.required',
+						'Received prop.required',
+						this.expand,
+					)
+
+				return 'valid'
+			},
+			pass: passDefault && passNoRequired,
+		}
+	},
+	toValidate(prop: PropOptions<unknown, boolean>, ...cases: unknown[]) {
+		const { validator } = prop
+
+		if (validator === undefined)
+			return {
+				message: () => 'Expected prop.validator to exist',
+				pass: false,
+			}
+
+		const successfulCases = cases.filter((c) => validator(c))
+		const failedCases = cases.filter((c) => !validator(c))
+
+		const wrongCases = this.isNot ? successfulCases : failedCases
```

Comment:
IMO, this overcomplicates just passing `success` and `failure` scenarios`. 
> the upside, I see, is that it's done like a test library would implement this, since at the end of the day, the API that's exposed will be chainable, thus, human-readable.

doesn't block anything, just don't see how this was a better refactor in terms of readability for the implementation itself, not for the usage of the end mutated API.

even more recognizable in the `pass` condition below https://github.com/3YOURMIND/kotti/pull/504/commits/83b368ff092c18779027c96cd9f717350e6bb9c9?diff=unified&w=1#diff-e0075d751357fe7b9821160bc2f8830ed316ea5140182506249baa0b298a59d3R120

Reply:
I initially had this entire function duplicated because using `.not` without switching between success and failure basically makes it wrong as it’s based on `all tests` or `no tests` but it flags `some tests` as well.

So, I basically added these 3 checks here to avoid having it duplicated entirely if that makes sense